### ISSUE TITLE
Big snap-to minimum jump - Hands move a lot when lock ratio gets checked

### ISSUE THREAD

**Issue by terracoda** [phetsims/ratio-and-proportion#257]
I think we are still working on how best to handle "at challenge" spots near the bottom of the screen, so this issue might naturally fix itself.

**To reproduce:**
1. Set My Challenge to 1:3
2. Use shift key to put Left hand at 1/3 and right hand at 1
    - The screen should be green & the hands should be at challenge ratio.
3. Go check the Lock Ratio checkbox
    - The hands jump and snap to 0.5 & 1.5

It's kind of a big hop & it's even a bigger hop with 1:7

I am just wondering if the snapping behaviour that snaps to the _lowest at challenge spot_ can be based on the smallest positive step with shift key which I think it 0.1, and then the snapping might be way less noticeable.

**Before and after ratio is locked with a 1:3 challenge**
![13-before-ratio-locked](https://user-images.githubusercontent.com/1902116/100388648-c3a0d800-3005-11eb-957a-2a18cb4e27f1.png)
![13-after-ratio-locked](https://user-images.githubusercontent.com/1902116/100388651-c56a9b80-3005-11eb-9fd3-d74be393ac57.png)

**Before and after ratio is locked with a 1:7 challenge**
![17-before-ratio-locked](https://user-images.githubusercontent.com/1902116/100388493-568d4280-3005-11eb-8449-ff37a6dff6ff.png)
![17-after-ratio-locked](https://user-images.githubusercontent.com/1902116/100388499-5c832380-3005-11eb-8459-70deecf9c89f.png)

**Comment by terracoda** [comment]
@emily-phet and @zepumph, is this a known problem?

**Comment by terracoda** [comment]
@emily-phet and @zepumph, is this a known problem?

**Comment by BLFiedler** [comment]
Is this only happening with the left hand below our minimum value for getting the ratio? Shouldn't it not be possible to get a perfect ratio below that so the lock ratio should not be enabled?

**Comment by terracoda** [comment]
I am putting the left hand below 0.5, and I am able to get the hands to "at challenge ratio" with the shift keys with the left hand below 0.5.

I am wondering if the minimum position for the left hand should be 0.1 (the shift key step size) rather than 0.5.

**Comment by terracoda** [comment]
Let me rephrase that...
I am wondering if the minimum position for the left hand (0.1) should guide the minimum possible at ratio spot?

**Comment by terracoda** [comment]
This seems to be as low as I can go.
![Screen Shot 2020-11-27 at 12 52 44 PM](https://user-images.githubusercontent.com/1902116/100469020-16cf6500-30b0-11eb-8945-c0215cc0b7b0.png)

I am not sure if it is better to prevent going that low or to allow going that low. It seems to work fine with the shift key.

The described positions are very rough, even with tick marks on. The value of the left hand which is in real terms 0.25, is described as "around tick mark 0.5".

For the record, I cannot get ratio with the right hand below 0.5.

**Comment by zepumph** [comment]
@BLFiedler, this is not below that threshold (which is .21 of a tick mark).

This is technically a known problem, in which the snapping will be larger lower down because the fitness algorithm is, by design, position independent. 

We could swap the snapping so that it was based off of the right side, but this is not really a solution, as you could still reproduce the issue with the inverse ratio (2:1). Or like @terracoda said, base the snapping off of the larger value so the smaller one always snaps. 

I personally don't feel like it is a big problem. The purpose of the snapping is to align the model with what the user is seeing/expecting. When the dark green/"at challenge ratio" occurs, the user thinks they are in proportion, even though the model isn't technically at perfect proportion. Thus, I feel like the current behavior is acceptable, and not too confusing. Furthermore, the positional description for each hand is vague enough that I doubt that someone will keep around these small regions, since, like @terracoda mentioned in https://github.com/phetsims/ratio-and-proportion/issues/257#issuecomment-734912186, there isn't much to gain out of a description like "left around tick mark .5, right on tick mark .5".

**Comment by zepumph** [comment]
>Or like @terracoda said, base the snapping off of the larger value so the smaller one always snaps.

That seemed like a really nice way to handle this, so that the larger value was never getting snapped, it may look something like this patch, but there was a reentry bug when one of the values was below .5 tick marks (could be unrelated, but I need to investigate).

```diff
Index: js/common/model/RAPRatio.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/RAPRatio.js	(revision 3e123b5f289fecbeb1f68efa7543a725a73946a2)
+++ js/common/model/RAPRatio.js	(date 1606762555632)
@@ -151,14 +151,22 @@
 
     // Alter the antecedent to match the target ratio
     const currentRatioTuple = this.ratioTupleProperty.value;
-    currentRatioTuple.antecedent = targetRatio * currentRatioTuple.consequent;
+
+
+    if ( currentRatioTuple.antecedent > currentRatioTuple.consequent ) {
+      currentRatioTuple.antecedent = targetRatio * currentRatioTuple.consequent;
+    }
+    else {
+      currentRatioTuple.consequent = currentRatioTuple.antecedent / targetRatio;
+    }
+
+    // Then clamp to be within the currently enabled range.
+    const newRatioTuple = this.clampRatioTupleValuesInRange( currentRatioTuple, targetRatio );
+    assert && assert( newRatioTuple !== currentRatioTuple, '' );
 
     // Make sure that the lock ratio listener won't try to mutate the new RAPRatioTuple
     this.lockRatioListenerEnabled = false;
-
-    // Then clamp to be within the currently enabled range.
-    this.ratioTupleProperty.value = this.clampRatioTupleValuesInRange( currentRatioTuple, targetRatio );
-
+    this.ratioTupleProperty.value = newRatioTuple;
     this.lockRatioListenerEnabled = true;
   }
 
```

**Comment by zepumph** [comment]
Ok. I went ahead with the commit. @terracoda now the smaller value with snap, so hopefully it will seem much less dramatic. Does this solve the problem to you?

**Comment by terracoda** [comment]
Ok, I'll check it out. My main concern was, with such a large jump, I felt like I needed to describe the new values.

**Comment by terracoda** [comment]
@zepumph, I don't see why we have to move the hands at all when Lock Ratio is turned on. The hands are already in ratio.
I just tested and there still seems to be a big jump. I'm not sure why I can get to "at ratio" with the left hand below 0.5, but I can't do so when ratio is locked? But maybe this issue does not have to be solved immediately, but maybe after the upcoming pre-dev dev test.

![1-7-before-lock-ratio](https://user-images.githubusercontent.com/1902116/100900936-efa4d900-349d-11eb-9a4a-e9a1d50a1ad3.png)

![1-after-lock-ratio](https://user-images.githubusercontent.com/1902116/100900971-f5022380-349d-11eb-9429-39665aa385af.png)

**Comment by terracoda** [comment]
In https://github.com/phetsims/ratio-and-proportion/issues/257#issue-751812436, the starting 1:7 "at ratio" position is different (values are higher), but the after lock ratio position is the same 0.5:3.5. Is that because with a 1:7 ratio, that's the first "even" tick representation?

**Comment by terracoda** [comment]
With a ratio of 1:3, the jump is less, of course.
Before I lock ratio, I can get the Left Hand to 2 shift-key steps below Tick Mark 0.5 (3 shift-key steps above zero) and move the Right Hand to 1 shift-key step below Tick Mark 3 and get the hands at ratio. This seems to mathematically correct, roughly 0.3:0.9. 

Because it seems mathematically correct to me, I don't see why we have to re-calculate once Lock Ratio is checked, but maybe as you say, it's not actually mathematically correct. 

**1:3 before Lock Ratio is checked**
hands at 0.3:0.9 (or there abouts)
![1-3-before-lock-ratio](https://user-images.githubusercontent.com/1902116/100903636-b752ca00-34a0-11eb-9cec-d1ed073fd378.png)

**1:3 after Lock Ratio is checked** 
hands at 0.5:1.5
![1-3-after-lock-ratio](https://user-images.githubusercontent.com/1902116/100903867-f3862a80-34a0-11eb-9578-ddef1c8fd50b.png)

Again, maybe this is about needing "even" values with regard to the actual Tick Mark values? Not sure?
And again, I'll say maybe this issue does not have to be solved immediately, but maybe after the upcoming pre-dev dev test.

My comments document the behaviour after your update.

**Comment by terracoda** [comment]
I am wondering is the "fitness" you have built in is based on the step-size of 0.1 Tick Marks or on some other value? And honestly, I don't know the math, so maybe that isn't even a good way to think about the fitness adjustment.

**Comment by terracoda** [comment]
@emily-phet and @zepumph, should we look at this issue again after the pre-dev dev-test is completed?

**Comment by zepumph** [comment]
Oh! I figured out the issue. We said at one point that the range of the locked ratio is between .5 and 10, so not only is the ratio snapping to the exact target, it is also getting moved up. @terracoda, I am not able to reproduce this though because I keep running into #258.

> I don't see why we have to move the hands at all when Lock Ratio is turned on. The hands are already in ratio.

Let me try to sum up an explanation of the code, which will likely be review.

"at ratio" is a description output which is a range of fitness. We need snapping to occur when ratio is locked because minor difference between the exact ratio and the state of the hands can be multiplied out to have large effects, see https://github.com/phetsims/ratio-and-proportion/issues/146.

I don't think it is very helpful to be using description position to describe this issue, because it is so vague. Instead can you reproduce a case using `?log` and copy over that model data for the "before locking" state and the after too? I think we will find that the left snaps up to .5, which is why the right snaps so far. Perhaps we can relax this line of code:

https://github.com/phetsims/ratio-and-proportion/blob/0daa8861933db285564928c26725f461649e3872/js/common/model/RAPRatio.js#L105-L107

The above says that when locked, the min value becomes .5.

**Comment by terracoda** [comment]
I think what you explain above is the issue. The left hand is indeed snapping to 0.5.
I'll get the numbers with the ?log to help us.

**Comment by terracoda** [comment]
I think the latest numbers are at the bottom of the log.
**Hands at ratio before lock ratio**
![Screen Shot 2020-12-02 at 4 40 09 PM](https://user-images.githubusercontent.com/1902116/100925976-4cfc5280-34bd-11eb-8966-0a5563ffe3b1.png)

**Hands at ratio after checking the lock ratio checkbox**
![Screen Shot 2020-12-02 at 4 40 43 PM](https://user-images.githubusercontent.com/1902116/100926011-54bbf700-34bd-11eb-9bc9-76670bfcfb15.png)

**Comment by terracoda** [comment]
Those look off. Here we go again:

**Before Lock Ratio**
![Screen Shot 2020-12-02 at 4 48 38 PM](https://user-images.githubusercontent.com/1902116/100926696-69e55580-34be-11eb-90b3-fcff4689a836.png)


**After Lock Ratio**
![Screen Shot 2020-12-02 at 4 49 04 PM](https://user-images.githubusercontent.com/1902116/100926726-749fea80-34be-11eb-9137-512543f994a8.png)

**Comment by terracoda** [comment]
In the first log, I think I resized the browser, causing issue #209. I think that's why the visuals look incorrect.

**Comment by terracoda** [comment]
Hopefully, the log values in https://github.com/phetsims/ratio-and-proportion/issues/257#issuecomment-737472498 are helpful.

**Comment by zepumph** [comment]
Thanks @terracoda, it is nice to confirm this. I appreciate the logs. This min value was determined as behavior that we wanted in https://github.com/phetsims/ratio-and-proportion/issues/92#issuecomment-665175449. We should bring this up at a meeting to see how we should proceed. 

What to do?
1. Leave it as is
2. Change the range of snapping
3. When locking, don't necessarily snap, but only let the user move towards .05, and then don't let them go back. I don't know how hard this would be to implement.

**Comment by terracoda** [comment]
@zepumph, thanks for the list. I was also thinking about the same 3 options, and especially curious about numbers 2 and 3.

**Comment by zepumph** [comment]
Today we talked mostly about why this range end of .05 is different from the NO_SUCCESS threshold in the model (.021). We decided that it should be the same, so I will change this .05 to follow that constant from the RAPModel.

**Comment by zepumph** [comment]
I thought this patch would work, but for some reason, the model won't respect the lock-ratio bounds when the range min is the same as the no-success threshold. I found that when I make the ratio threshold Number.EPSILON larger though, it works. I don't want to commit that workaround, I want to figure out why.

```js
Index: js/common/RAPConstants.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/RAPConstants.js b/js/common/RAPConstants.js
--- a/js/common/RAPConstants.js	(revision 77726495d143a3b7920e7daac41d203b19e42ddb)
+++ b/js/common/RAPConstants.js	(date 1608143977722)
@@ -30,7 +30,12 @@
   TOTAL_RATIO_TERM_VALUE_RANGE: new Range( 0, 1 ),
 
   // Consistent way to fix numbers. This should only be used in the view for comparison and display, not in the model, see https://github.com/phetsims/ratio-and-proportion/issues/243
-  toFixed: x => Utils.toFixedNumber( x, 6 )
+  toFixed: x => Utils.toFixedNumber( x, 6 ),
+
+  // REVIEW: Seems to have a typo in the name and the value in the description doesn't match the value assigned.
+  // The value in which when either the antecedent or consequent is less than this, the ratio cannot be "in proportion".
+  // Add .001 to support two keyboard nav motions above 0 (counting the min range being >0).
+  NO_SUCCUSS_VALUE_THRESHOLD: .01
 };
 
 ratioAndProportion.register( 'RAPConstants', RAPConstants );
Index: js/common/model/RAPModel.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/RAPModel.js b/js/common/model/RAPModel.js
--- a/js/common/model/RAPModel.js	(revision 77726495d143a3b7920e7daac41d203b19e42ddb)
+++ b/js/common/model/RAPModel.js	(date 1608143977722)
@@ -17,11 +17,6 @@
 // constant to help achieve feedback in 40% of the visual screen height (2 default tick marks).
 const FITNESS_TOLERANCE_FACTOR = 0.5;
 
-//REVIEW: Seems to have a typo in the name and the value in the description doesn't match the value assigned.
-// The value in which when either the antecedent or consequent is less than this, the ratio cannot be "in proportion".
-// Add .001 to support two keyboard nav motions above 0 (counting the min range being >0).
-const NO_SUCCUSS_VALUE_THRESHOLD = .01;
-
 const TOTAL_RANGE = RAPConstants.TOTAL_RATIO_TERM_VALUE_RANGE;
 
 class RAPModel {
@@ -150,7 +145,8 @@
    * @returns {boolean}
    */
   valuesTooSmallForInProportion() {
-    return this.ratio.antecedentProperty.value <= NO_SUCCUSS_VALUE_THRESHOLD || this.ratio.consequentProperty.value <= NO_SUCCUSS_VALUE_THRESHOLD;
+    return this.ratio.antecedentProperty.value <= RAPConstants.NO_SUCCUSS_VALUE_THRESHOLD ||
+           this.ratio.consequentProperty.value <= RAPConstants.NO_SUCCUSS_VALUE_THRESHOLD;
   }
 
   /**
Index: js/common/model/RAPRatio.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/common/model/RAPRatio.js b/js/common/model/RAPRatio.js
--- a/js/common/model/RAPRatio.js	(revision 77726495d143a3b7920e7daac41d203b19e42ddb)
+++ b/js/common/model/RAPRatio.js	(date 1608144033625)
@@ -21,7 +21,7 @@
 const VELOCITY_THRESHOLD = .01;
 
 const DEFAULT_TERM_VALUE_RANGE = RAPConstants.TOTAL_RATIO_TERM_VALUE_RANGE;
-const LOCK_RATIO_RANGE_MIN = .05;
+const LOCK_RATIO_RANGE_MIN = RAPConstants.NO_SUCCUSS_VALUE_THRESHOLD; // Use the same value as the no-success region threshold
 
 class RAPRatio {
   constructor() {

```

**Comment by zepumph** [comment]
Ok. I finally understand the problem. Conceptually having these thresholds be the same don't make sense. 

We want the number in which you cannot have successful fitness (so fitness must be <1), be the same value that a locked ratio can get to. The definition of a locked ratio is a ratio in proportion. Thus, I now feel like my Number.EPSILON workaround is making more and more sense.

**Comment by zepumph** [comment]
@terracoda, after the above commit, the range for the lock ratio goes to just shy of the no-success threshold. So right now, from 0-.1 of a tickmark (.01 in the model), you can't be in-proportion. Then from `(.01,1]` (.01 exclusive), you can move the locked ratio around. Please let me know if this is good to you.

**Comment by terracoda** [comment]
There is no jumping when I lock the ratio. I tried several different ratios: 
- 1:2
- 2:7
- 1:10
- 2:10
- 1:8

I moved hands into the lowest possible place I could get. The left hand would always have to be 2 shift-key steps up from the bottom. For example with ratio 1:10, the lowest I could physically go without Lock Ratio checked was with the Right Hand on Tick Mark 2 and the Left hand 2 shift-key steps above Tick Mark 0.

I locked the Ratio, and the hands did not jump!

Continuing with ratio locked, once I got back to the hands, I could move them down further with ratio locked.

I realize now, I think that the left hand doesn't always move a full shift-key step depending on how close to proportion the hands actually are. This seems to only be an issue down near zero.

There is no more jumping, so this issue is resolved.
I think it is working really well now.

I have one remaining question though, is it ok that I can explore lower with ratio locked than when ratio is not locked? Is that expected?

I'll close this issue since you have addressed the issue, and I will open a new one with my question. Hopefully, we can quickly close that one, too.
